### PR TITLE
GIX-1541: Get user location if needed

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

We want to restrict participation of users from certain countries to participate in SNS swap.

In this PR: Load the location of the user when needed.

Note: Not fully functional yet because the backend is missing the necessary field.

# Changes

* New project util `userCountryIsNeeded`.
* Use new project util in ProjectDetail to call the service `loadUserCountry`.

# Tests

* Add tests for new project util
* Add test in ProjectDetail for new functionality.

### CODE REVIEW

File: frontend/src/lib/pages/ProjectDetail.svelte [hunk 1]
```diff
@@ -161,6 +163,16 @@
...
   }
 
+  let shouldLoadUserCountry = false;
+  $: shouldLoadUserCountry = userCountryIsNeeded({
+    summary: $projectDetailStore?.summary,
+    swapCommitment: $projectDetailStore?.swapCommitment,
+    loggedIn: $authSignedInStore,
+  });
+  $: if (shouldLoadUserCountry) {
+    loadUserCountry();
```
Comment (lmuntaner) [inline frontend/src/lib/pages/ProjectDetail.svelte:173]: This willl not be called until `deny_list` is added,

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts [hunk 2]
```diff
@@ -287,6 +295,19 @@ sale_buyer_count ${saleBuyerCount} 1677707139456
         ).toMatch(formatToken({ value: userCommitment }));
       });
 
+      // TODO: GIX-1541 use this test to show that button is disabled if user is in a country that is not allowed to participate
+      it("should NOT load user's country", async () => {
```
Comment (lmuntaner) [inline frontend/src/tests/lib/pages/ProjectDetail.spec.ts:299]: I want to keep this test to make sure the service is not yet used.

File: frontend/src/lib/utils/projects.utils.ts [hunk 3]
```diff
@@ -158,6 +159,34 @@ export const canUserParticipateToSwap = ({
   );
 };
 
+/**
+ * Returns whether the location of the user is needed to participate to the swap.
+ *
+ * Some projects might restrict participation in the swap to users from specific countries.
+ *
+ * Yet, we don't need to get the location of all users, only those who can to participate to the swap:
```
Comment (dskloetd) [inline frontend/src/lib/utils/projects.utils.ts:167]: "only those who can to participate to the swap"
Is there a typo there? I'm not sure how to parse this.

File: frontend/src/lib/utils/projects.utils.ts [hunk 4]
```diff
@@ -158,6 +159,34 @@ export const canUserParticipateToSwap = ({
...
+  swapCommitment: SnsSwapCommitment | undefined | null;
+  loggedIn: boolean;
+}): boolean =>
+  canUserParticipateToSwap({ summary, swapCommitment }) &&
+  loggedIn &&
+  nonNullish(summary) &&
+  // TODO: GIX-1541 Remove unncessary checking and casting of deny_list
+  "deny_list" in summary.swap &&
+  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+  (summary.swap.deny_list as any).length > 0;
```
Comment (dskloetd) [inline frontend/src/lib/utils/projects.utils.ts:188]: Can you cast it to `string[]` instead of to `any`?

File: frontend/src/tests/lib/utils/projects.utils.spec.ts [hunk 5]
```diff
@@ -272,6 +273,111 @@ describe("project-utils", () => {
...
 
+  describe("userCountryIsNeeded", () => {
+    it("country not needed if no summary or swap information", () => {
+      expect(
+        userCountryIsNeeded({
+          summary: undefined,
+          swapCommitment: undefined,
+          loggedIn: true,
+        })
+      ).toBeFalsy();
```
Comment (dskloetd) [inline frontend/src/tests/lib/utils/projects.utils.spec.ts:284]: Just curious, why `toBeFalsy` instead of the stricter `toBe(false)`? Do you expect it to potentially return null or undefined?

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts (hunk 2 above)
Comment (dskloetd) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts:299]: Is it possible to also have a test where the user's country is loaded?
Otherwise it's hard to be confident that this test sets up the right conditions where the country would be loaded. If all you test is that something doesn't happen, the test also passes if the setup is completely broken and nothing happens.

File: frontend/src/tests/lib/utils/projects.utils.spec.ts [hunk 6]
```diff
@@ -272,6 +273,111 @@ describe("project-utils", () => {
...
+      expect(
+        userCountryIsNeeded({
+          summary: summaryForLifecycle(SnsSwapLifecycle.Aborted),
+          swapCommitment: mockSwapCommitment,
+          loggedIn: true,
+        })
+      ).toBeFalsy();
+    });
+
+    // TODO: GIX-1541 Add a project with deny_list and test that returns true
```
Comment (dskloetd) [inline frontend/src/tests/lib/utils/projects.utils.spec.ts:342]: If you do some casting, similar to the tested code, you could have a test that returns true, right?

File: frontend/src/lib/utils/projects.utils.ts [hunk 7]
```diff
@@ -158,6 +159,34 @@ export const canUserParticipateToSwap = ({
...
+  loggedIn,
+}: {
+  summary: SnsSummary | undefined | null;
+  swapCommitment: SnsSwapCommitment | undefined | null;
+  loggedIn: boolean;
+}): boolean =>
+  canUserParticipateToSwap({ summary, swapCommitment }) &&
+  loggedIn &&
+  nonNullish(summary) &&
+  // TODO: GIX-1541 Remove unncessary checking and casting of deny_list
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply frontend/src/lib/utils/projects.utils.ts:185]: Looking at [SnsSummary](https://github.com/dfinity/nns-dapp/blob/main/frontend/src/lib/types/sns.ts#L42), am I right to assume that the denied countries list will most likely be on the [SnsParams](https://github.com/dfinity/nns-dapp/blob/main/frontend/src/lib/types/sns.ts#L37) in the `SnsSummarySwap`?

File: frontend/src/lib/utils/projects.utils.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:167]: We only need the location of the users that can participate in the swap.

Does that help?

Is "swap" the problem here?

File: frontend/src/lib/utils/projects.utils.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts:188]: It won't be `string[]`. It's not yet clear the format.

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts:284]: No, I was thinking on making it easier to read and it's what we have also on another util.

No strong opinion.

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts]: good idea, done!

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts]: done

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts]: I used the "getter" idea and I mock it when I want to test a different value than empty string.

Let me know what you think.

File: frontend/src/lib/utils/projects.utils.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/utils/projects.utils.ts]: I used the getter idea.

**Review by dskloetd — COMMENTED** [review]
GitHub won't let me submit my review without typing something here.

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts [hunk 8]
```diff
@@ -287,6 +296,32 @@ sale_buyer_count ${saleBuyerCount} 1677707139456
...
+          rootCanisterId: Principal.fromText(rootCanisterId),
+          myCommitment: undefined,
+        } as SnsSwapCommitment);
+
+        render(ProjectDetail, props);
+
+        expect(locationApi.queryUserCountryLocation).not.toBeCalled();
+      });
+
+      it("should load user's country if no deny list", async () => {
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts:311]: yes, sorry, thanks!

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts [hunk 9]
```diff
@@ -287,6 +296,32 @@ sale_buyer_count ${saleBuyerCount} 1677707139456
         ).toMatch(formatToken({ value: userCommitment }));
       });
 
+      it("should not load user's country if no deny list", async () => {
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts:299]: Which tests together?

* should not load user's country if no deny list
* should load user's country if non-empty deny list

or

* should load user's country if non-empty deny list
* should participate without user interaction if there is an open ticket.

This is already nested inside three describes, that's why I wasn't sure about nesting one further.

File: frontend/src/lib/getters/sns-summary.ts [hunk 10]
```diff
@@ -0,0 +1,4 @@
+import type { SnsSummary } from "$lib/types/sns";
+
+// TODO: GIX-1545 Implement this function
+export const getDenyList = (_summary: SnsSummary) => [];
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/getters/sns-summary.ts:4]: sure

File: frontend/src/tests/lib/utils/projects.utils.spec.ts [hunk 11]
```diff
@@ -272,6 +274,115 @@ describe("project-utils", () => {
...
+          summary: summaryForLifecycle(SnsSwapLifecycle.Aborted),
+          swapCommitment: mockSwapCommitment,
+          loggedIn: true,
+        })
+      ).toBe(false);
+    });
+
+    it("country is needed", () => {
+      // TODO: TODO: GIX-1545 Remove mock and create a summary with deny list
+      jest.spyOn(summaryGetters, "getDenyList").mockReturnValue(["US"]);
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts:348]: Do you mean to add two describe?

The others require the deny list to be empty. If I mock it for them returning something they will fail.

I'm not sure what you mean.

Since you brought up this pattern of putting the mock in a beforeEach before. Can you describe an objective rule of when it should be there and when not?

It would be nice to be consistent in all tests also.

[the remark this answers is not in the document]
Comment (dskloetd) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts]: > The others require the deny list to be empty.

Do they? I think in every other test it will not even call `getDeniedCountries()`.
But if they accidentally do call `getDeniedCountries()` we don't want *that* to be the reason it returns false.
For example in `"country not needed if sale is not open"` we want the fact that the sale is not open to be the reason it returns false. So every other potential reason should be set to true.

I also now realized that a test seems to be missing where every other reason is true but the getDeniedCountries() list is not present. Or am I over looking it?

Basically you are testing a function like this:
`getCondition(a, b, c, d) => a && b && c && d`
Then I would expect tests where
* a, b and c are true but d is false
* a, b, and d are true but c is false
* a, c and d are true but d is false, and
* b, c and d are true but a is false

So in the beforeEach I would set a, b, c and d all to true and then in each of the tests set 1 thing to false.
Does that make sense?

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts:299]: Ah, because they both share the setup for `getOpenTicket` and `querySnsSwapCommitment`. Got it.

File: frontend/src/tests/lib/utils/projects.utils.spec.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/utils/projects.utils.spec.ts]: Yes, that makes sense.

One thing I'm not sure is overriding mocks. I don't like the idea that a specific test overrides the setup.

But this won't apply here. So all good.

File: frontend/src/tests/lib/pages/ProjectDetail.spec.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/ProjectDetail.spec.ts]: done

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

**Review by dskloetd — APPROVED** [review]
Thanks!
